src/lib/tls/tls_session_manager.cpp

```diff
@@ -0,0 +1,184 @@
+/**
+ * TLS Session Manager
+ * (C) 2011 Jack Lloyd
+ *     2023 René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#include <botan/tls_session_manager.h>
+
+#include <botan/tls_callbacks.h>
+#include <botan/tls_policy.h>
+#include <botan/rng.h>
+
+namespace Botan::TLS {
+
+Session_Manager::Session_Manager(RandomNumberGenerator& rng)
+   : m_rng(rng) {}
+
+std::optional<Session_Handle> Session_Manager::establish(const Session& session,
+            std::optional<Session_ID> id,
+            bool tls12_no_ticket)
+{
+   // By default, the session manager does not emit session tickets anyway
+   BOTAN_UNUSED(tls12_no_ticket);
+   BOTAN_ASSERT(session.side() == Connection_Side::SERVER,
+                "Client tried to establish a session");
+
+   Session_Handle handle(id.value_or(m_rng.random_vec<Session_ID>(32)));
+   store(session, handle);
+   return handle;
+}
+
+std::optional<Session> Session_Manager::retrieve(const Session_Handle& handle,
+                                                 Callbacks& callbacks,
+                                                 const Policy& policy)
+   {
+   BOTAN_UNUSED(callbacks, policy);
+   auto session = retrieve_one(handle);
+   if(!session.has_value())
+      return std::nullopt;
+
+   // A value of '0' means: No policy restrictions.
+   const std::chrono::seconds policy_lifetime =
+      (policy.session_ticket_lifetime() > 0)
+         ? std::chrono::seconds(policy.session_ticket_lifetime())
+         : std::chrono::seconds::max();
+
+   // RFC 5077 3.3 -- "Old Session Tickets"
+   //    A server MAY treat a ticket as valid for a shorter or longer period of
+   //    time than what is stated in the ticket_lifetime_hint.
+   //
+   // RFC 5246 F.1.4 -- TLS 1.2
+   //    If either party suspects that the session may have been compromised, or
+   //    that certificates may have expired or been revoked, it should force a
+   //    full handshake.  An upper limit of 24 hours is suggested for session ID
+   //    lifetimes.
+   //
+   // RFC 8446 4.6.1 -- TLS 1.3
+   //    A server MAY treat a ticket as valid for a shorter period of time than
+   //    what is stated in the ticket_lifetime.
+   //
+   // Note: This disregards what is stored in the session (e.g. "lifetime_hint")
+   //       and only takes the local policy into account. The lifetime stored in
+   //       the sessions was taken from the same policy anyways and changes by
+   //       the application should have an immediate effect.
+   const auto ticket_age = callbacks.tls_current_timestamp() - session->start_time();
+   const bool expired = ticket_age > policy_lifetime;
```

Comment:
This applies the same expiry filter to both tickets and TLS 1.2 session IDs. Previously, there were two lifetimes:
* `Policy::session_ticket_lifetime()` and
* `Session_Manager_Memory/SQL(..., session_lifetime)`.

TLS 1.3 doesn't make this distinction anyway, but we could introduce a `Policy::tls12_session_id_lifetime()` to allow the same flexibility as before.

Note that previously derived `Session_Manager` classes each needed to implement the expiry logic. That is still possible (by overriding `::retrieve()` and `::find()`) but not necessary anymore. For applications to use the default implementation, they would just need to override the "internal storage methods" `::retrieve_one()` and `::find_all()`.